src/base/utils/environment_p.cpp

```diff
@@ -0,0 +1,292 @@
+{
+#ifdef Q_OS_WIN
+    return SettingsPtr(new QSettings(QSettings::IniFormat, QSettings::UserScope, configurationName(), name));
+#else
+    return SettingsPtr(new QSettings(configurationName(), name));
+#endif
+}
+
+const QString Utils::Envrironment::impl::SpecialFoldersImplCustom::cacheDirName = QLatin1String("cache");
+const QString Utils::Envrironment::impl::SpecialFoldersImplCustom::configDirName = QLatin1String("config");
+const QString Utils::Envrironment::impl::SpecialFoldersImplCustom::dataDirName = QLatin1String("data");
+const QString Utils::Envrironment::impl::SpecialFoldersImplCustom::downloadsDirName = QLatin1String("downloads");
+
+Utils::Envrironment::impl::SpecialFoldersImplCustom::SpecialFoldersImplCustom(const QString &rootPath, bool create, const QString &configurationName)
+    : SpecialFoldersImpl(configurationName)
+    , m_rootDirectory {QDir(rootPath).absoluteFilePath(this->configurationName())}
+{
+    if (create) {
+        if (!m_rootDirectory.mkpath(cacheDirName))
+            qFatal("Can not create '%s' subdirectory under the profile directory '%s'",
+                   qPrintable(cacheDirName), qPrintable(m_rootDirectory.canonicalPath()));
+        if (!m_rootDirectory.mkpath(configDirName))
+            qFatal("Can not create '%s' subdirectory under the profile directory '%s'",
+                   qPrintable(configDirName), qPrintable(m_rootDirectory.canonicalPath()));
+        if (!m_rootDirectory.mkpath(dataDirName))
+            qFatal("Can not create '%s' subdirectory under the profile directory '%s'",
+                   qPrintable(dataDirName), qPrintable(m_rootDirectory.canonicalPath()));
+        if (!m_rootDirectory.mkpath(downloadsDirName))
+            qFatal("Can not create '%s' subdirectory under the profile directory '%s'",
+                   qPrintable(downloadsDirName), qPrintable(m_rootDirectory.canonicalPath()));
+    }
+}
+
+QString Utils::Envrironment::impl::SpecialFoldersImplCustom::cacheLocation() const
+{
+    return m_rootDirectory.absoluteFilePath(cacheDirName);
+}
+
+QString Utils::Envrironment::impl::SpecialFoldersImplCustom::configLocation() const
+{
+    return m_rootDirectory.absoluteFilePath(configDirName);
+}
+
+QString Utils::Envrironment::impl::SpecialFoldersImplCustom::dataLocation() const
+{
+    return m_rootDirectory.absoluteFilePath(dataDirName);
+}
+
+QString Utils::Envrironment::impl::SpecialFoldersImplCustom::downloadLocation() const
+{
+    return m_rootDirectory.absoluteFilePath(downloadsDirName);
+}
+
+Utils::Envrironment::SettingsPtr Utils::Envrironment::impl::SpecialFoldersImplCustom::applicationSettings(const QString &name) const
+{
+    // here we force QSettings::IniFormat format because we need it to be portable across platforms
+    return SettingsPtr(new QSettings(QDir(configLocation()).absoluteFilePath(name + QLatin1String(".conf")), QSettings::IniFormat));
```

Comment:
maybe just me, you could use a temporary QString to hold the QDir::absoluteFilePath(), will be easier to read.